File: problems/tier2-intermediate/010-ride-surge-pricing/boilerplate/go/part2/guided.go
package main

// Data class (given).
type PricingContext struct {
	baseFare float64
	availableDrivers int
	activeRideRequests int
	timeOfDay string
	weather string
}

type RideRequest struct {
	userId string
	pickup string
	dropoff string
	rideType string
}

// HINT: introduce an abstraction so new rules don't change existing code.

// HINT: start from what this must return, then work backwards to the state it needs.
func calculateSurge(ctx PricingContext) float64 {
	// TODO: write your solution
	return 0.0
}

// HINT: start from what this must return, then work backwards to the state it needs.
func calculateFare(req RideRequest, ctx PricingContext) float64 {
	// TODO: write your solution
	return 0.0
}

File: problems/tier2-intermediate/010-ride-surge-pricing/boilerplate/javascript/part2/guided.js
// Data class (given).
class PricingContext {
  constructor(baseFare, availableDrivers, activeRideRequests, timeOfDay, weather) {
    this.baseFare = baseFare;
    this.availableDrivers = availableDrivers;
    this.activeRideRequests = activeRideRequests;
    this.timeOfDay = timeOfDay;
    this.weather = weather;
  }
}

class RideRequest {
  constructor(userId, pickup, dropoff, rideType) {
    this.userId = userId;
    this.pickup = pickup;
    this.dropoff = dropoff;
    this.rideType = rideType;
  }
}

// HINT: introduce an abstraction so new rules don't change existing code.

// HINT: start from what this must return, then work backwards to the state it needs.
function calculateSurge(ctx) {
  // TODO: write your solution
  return null;
}

// HINT: start from what this must return, then work backwards to the state it needs.
function calculateFare(req, ctx) {
  // TODO: write your solution
  return null;
}

// ── Export everything the test runner needs (do not remove) ──
// If you add classes (e.g. strategy subclasses), add them here too.
module.exports = { PricingContext, RideRequest, calculateSurge, calculateFare };
